### PR TITLE
Added preserve_time parameter to move, copy, and mirror

### LINKED ISSUE #450: original timestamp (ctime, mtime) cannot survive a copy_fs or mirror
as long as there's actual "copy" operation from one fs to another, then those "touched" files will lost their original ctime or mtime, which would be updated to the time when that copy operation happened.
this is not expected feature. though calling it "critical bug" might be too harsh but still it's much better if it could be fixed.

### PR DESCRIPTION
**Pull request by atollk**
## Type of changes

- New feature

## Checklist

- [x] I've run the latest [black](https://github.com/ambv/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @PyFilesystem/maintainers may be pedantic in the code review.

## Description

Fixes #450 by adding a `preserve_time` parameter to all relevant functions. The parameter is `False` by default to retain previous behaviour. However, it could be argued that the default should be `True` to mimic more common behavior found in operating systems.

### CODE REVIEW

**Comment by atollk** [comment]
I wanted to add https://github.com/wolever/parameterized as a dependency because parameterization was done already anyway, just in a hacky self-implemented way. I don't actually get how to add test dependencies though?

**Comment by althonos** [comment]
If you want to add a dependency for the tests, add it to `tests/requirements.txt`, pinning the version with `~=` if possible. It will get installed by `tox` and by the CI runner.

**Comment by atollk** [comment]
@althonos From what I can tell, this doesn't work [here](https://ci.appveyor.com/project/willmcgugan/pyfilesystem2/build/job/d6v8sh072h8e27ni), as the tests are run without tox.

**Comment by althonos** [comment]
@atollk : that was not expected, I patched `appveyor.yml`. I need to refactor the CI on AppVeyor anyway, since we are currently not testing all supported versions or measuring coverage there.

**Comment by atollk** [comment]
I just saw that `FTPFS.setinfo` actually doesn't do anything, so I also added support for the MFMT command so that `preserve_time` works when copying to a FTPFS too.

File: fs/_bulk.py [hunk 1]
```diff
@@ -88,7 +97,7 @@ def __init__(self, num_workers=4):
     def start(self):
         """Start the workers."""
         if self.num_workers:
-            self.queue = Queue(maxsize=self.num_workers)
+            self.queue = Queue()
```
Comment (althonos) [inline fs/_bulk.py:100]: Why the change here?

File: fs/base.py [hunk 2]
```diff
@@ -417,16 +426,25 @@ def copy(self, src_path, dst_path, overwrite=False):
             with closing(self.open(src_path, "rb")) as read_file:
                 # FIXME(@althonos): typing complains because open return IO
                 self.upload(dst_path, read_file)  # type: ignore
+            copy_mtime(self, src_path, self, dst_path)
```
Comment (althonos) [inline fs/base.py:429]: Shouldn't it be:
```python
if preserve_time:
    copy_mtime(...)
```
here?

File: fs/base.py [hunk 3]
```diff
@@ -1133,6 +1155,7 @@ def move(self, src_path, dst_path, overwrite=False):
             with self.open(src_path, "rb") as read_file:
                 # FIXME(@althonos): typing complains because open return IO
                 self.upload(dst_path, read_file)  # type: ignore
+            copy_mtime(self, src_path, self, dst_path)
```
Comment (althonos) [inline fs/base.py:1158]: Same here:
```python
if preserve_time:
    copy_mtime(...)
```

Comment (althonos) [reply fs/base.py:1158]: Also, what happens if both resources have a syspath? There is an early return above, so I don't think the *mtime* is copied in that case.

File: fs/copy.py [hunk 4]
```diff
@@ -286,9 +334,10 @@ def dst():
     from ._bulk import Copier
 
     with src() as _src_fs, dst() as _dst_fs:
-        with _src_fs.lock(), _dst_fs.lock():
-            _thread_safe = is_thread_safe(_src_fs, _dst_fs)
-            with Copier(num_workers=workers if _thread_safe else 0) as copier:
+        _thread_safe = is_thread_safe(_src_fs, _dst_fs)
+        copier = Copier(num_workers=workers if _thread_safe else 0)
+        with copier:
+            with _src_fs.lock(), _dst_fs.lock():
```
Comment (althonos) [inline fs/copy.py:340]: Any reason to create the copier before locking instead of after like before?

File: fs/memoryfs.py [hunk 5]
```diff
@@ -444,7 +444,7 @@ def makedir(
                 parent_dir.set_entry(dir_name, new_dir)
             return self.opendir(path)
 
-    def move(self, src_path, dst_path, overwrite=False):
+    def move(self, src_path, dst_path, overwrite=False, preserve_time=False):
```
Comment (althonos) [inline fs/memoryfs.py:447]: `MemoryFS` actually supports created/modified times, so you should be able to get it to preserve times here too.

File: fs/wrapfs.py [hunk 6]
```diff
@@ -167,18 +167,18 @@ def makedir(
...
         dst_fs, _dst_path = self.delegate_path(dst_path)
         with unwrap_errors({_src_path: src_path, _dst_path: dst_path}):
             if not overwrite and dst_fs.exists(_dst_path):
                 raise errors.DestinationExists(_dst_path)
-            move_file(src_fs, _src_path, dst_fs, _dst_path)
+            move_file(src_fs, _src_path, dst_fs, _dst_path, preserve_time=preserve_time)
 
-    def movedir(self, src_path, dst_path, create=False):
-        # type: (Text, Text, bool) -> None
+    def movedir(self, src_path, dst_path, create=False, preserve_time=False):
```
Comment (althonos) [inline fs/wrapfs.py:180]: I think you're not using `preserve_time` in the subsequent `move_dir` call here.

**Review by althonos — CHANGES_REQUESTED** [review]
I'm not 100% satisfied with the `copy_mtime`  name, because we never refer to `mtime` anywhere else in the API (we use `modified` everywhere). Yet `copy_modified` does not convey the function usecase very well. `copy_modified_time` maybe? Might be a bit too long.

There are just a few minore issues here and there but overall this looks fine.

File: fs/base.py (hunk 3 above)
Comment (atollk) [reply fs/base.py:1158]: I think `os.rename` doesn't actually touch the mtime value, at least on UNIX and Windows. I added the check anyway, just to be safe.

File: fs/copy.py (hunk 4 above)
Comment (atollk) [reply fs/copy.py:340]: I had to change the logic of `Copier` because of some internal locking mechanics. But I just noticed that that change might actually have caused all workers to work sequentially, rather than in parallel. So I reverted most of my changes in `fs._bulk`. The modification times are now updated when closing the `Copier` instance, after all tasks have finished.

File: fs/_bulk.py (hunk 1 above)
Comment (atollk) [reply fs/_bulk.py:100]: see my other comment [here](https://github.com/PyFilesystem/pyfilesystem2/pull/463#discussion_r609513389)

**Review by althonos — APPROVED** [review]
LGTM, congrats!

**Comment by althonos** [comment]
I merged this one because I have already been too nit-picky here, but for the remaining PRs, would you be able to rebase them to `master` so that the commit history is still a bit organized?

**Comment by atollk** [comment]
@althonos Sure can do. But I'd also be happy to write a more thorough commit message for a squash merge, since I feel a good part of the commits in those PRs are just one-file fixes anyway. Or do you not do squashes in this repo?

**Comment by althonos** [comment]
The developer interaction was minimal here for some time so we just went with merges. I do squashes in other projects but for PyFilesystem I didn't wanna change what was there so branch/merges are fine. As long as they are properly rebased, the network is still fine: https://github.com/PyFilesystem/pyfilesystem2/network

**Comment by atollk** [comment]
@althonos In case you didn't see, I did a rebase in both https://github.com/PyFilesystem/pyfilesystem2/pull/462 and https://github.com/PyFilesystem/pyfilesystem2/pull/464.

**Comment by althonos** [comment]
Thanks, perfect, i'll have a look tomorrow !
